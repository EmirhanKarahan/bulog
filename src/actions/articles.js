import moment from "moment";
import { history } from "../routers/AppRouter";
import {
  addArticleFirebase,
  deleteImageFirebase,
  editArticleFirebase,
  getArticleByIdFirebase,
  getArticlesFirebase,
  removeArticleFirebase,
  uploadImageFirebase,
} from "../firebase/utils";

export const addArticle = (article) => {
  return {
    type: "ADD_ARTICLE",
    article,
  };
};

export const startAddArticle = (articleData = {}) => {
  return async (dispatch) => {
    const {
      title = "",
      subtitle = "",
      content = "",
      image = undefined,
      date = moment().unix(),
    } = articleData;

    const imageUrl = await uploadImageFirebase(image);
    const article = { title, subtitle, content, date, imageUrl };
    const ref = await addArticleFirebase(article);
    dispatch(addArticle({ id: ref.key, ...article }));
    history.push("/");
  };
};

export const editArticle = (id, updates) => ({
  type: "EDIT_ARTICLE",
  id,
  updates,
});

export const startEditArticle = (
  id,
  { title, subtitle, content, image, imageUrl }
) => {
  return async (dispatch) => {
    if (image) {
      await deleteImageFirebase(imageUrl);
      imageUrl = await uploadImageFirebase(image);
    }
    const updates = { title, subtitle, content, imageUrl };
    await editArticleFirebase(id, updates);
    dispatch(editArticle(id, updates));
    history.push(`/read/${id}`);
  };
};

export const removeArticle = ({ id } = {}) => ({
  type: "REMOVE_ARTICLE",
  id,
});

export const startRemoveArticle = ({ id } = {}) => {
  return async (dispatch, getState) => {
    dispatch(removeArticle({ id }));
    const snapshot = await getArticleByIdFirebase(id);
    const imageUrl = snapshot.val().imageUrl;
    await deleteImageFirebase(imageUrl);
    await removeArticleFirebase(id);
  };
};

export const setArticles = (articles) => ({
  type: "SET_ARTICLES",
  articles,
});

export const startSetArticles = () => {
  return async (dispatch, getState) => {
    const snapshot = await getArticlesFirebase();
    const articles = [];
    snapshot.forEach((childSnapshot) => {
      articles.push({ id: childSnapshot.key, ...childSnapshot.val() });
    });
    dispatch(setArticles(articles));
  };
};
