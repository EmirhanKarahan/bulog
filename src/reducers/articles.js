const articlesReducerDefaultState = [];

export default (state = articlesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_ARTICLE':
      return [
        ...state,
        action.article
      ];
    case 'REMOVE_ARTICLE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_ARTICLE':
      return state.map((article) => {
        if (article.id === action.id) {
          return {
            ...article,
            ...action.updates
          };
        } else {
          return article;
        };
      });
    case 'SET_ARTICLES':
      return action.articles;
    default:
      return state;
  }
};