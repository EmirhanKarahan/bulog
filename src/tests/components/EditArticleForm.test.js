import React from "react";
import { shallow } from "enzyme";
import EditArticleFormik from "../../components/EditArticleForm";
import articles from "../fixtures/articles"

test("should render EditArticleForm correctly", () => {
  const wrapper = shallow(<EditArticleFormik {...articles[0]}/>);
  expect(wrapper).toMatchSnapshot();
});