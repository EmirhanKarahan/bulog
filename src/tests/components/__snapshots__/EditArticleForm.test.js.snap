// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditArticleForm correctly 1`] = `
<Formik
  content="Cool article"
  date={0}
  handleSubmit={[Function]}
  id="1"
  imageUrl="https://firebasestorage.googleapis.com/1"
  initialValues={
    Object {
      "content": "Cool article",
      "image": undefined,
      "imageUrl": "https://firebasestorage.googleapis.com/1",
      "subtitle": "First Article's Subtitle",
      "title": "First Article",
    }
  }
  onSubmit={[Function]}
  subtitle="First Article's Subtitle"
  title="First Article"
  validationSchema={[Function]}
>
  <Component />
</Formik>
`;
